src/server/data_export.py

```diff
@@ -0,0 +1,59 @@
+#
+# Data export handlers
+#
+
+#
+# Data exporters first collect data via their assembler function,
+# then pass that data through the formatter function before output.
+# PageCache is updated before export() is called and can be assumed accurate.
+#
+# name should be unique and acts as an identifier
+# label becomes visible in the UI and becomes translatable
+# formatter(data) should be used for final handling of file format
+# assembler(Database, PageCache) collects relevant data from timer
+#   before formatting.
+#
+
+import logging
+from Plugins import Plugins
+
+class dataExportManager():
+    exporters = {}
+
+    def __init__(self):
+        self.readPlugins()
+
+    def readPlugins(self):
+        exporter_plugins = Plugins(prefix='data_export')
+        exporter_plugins.discover()
+
+        for exporter in exporter_plugins:
+            self.registerExporter(exporter)
+
+    def registerExporter(self, exporter):
+        if hasattr(exporter, 'name'):
+            if exporter.name in self.exporters:
+                logger.warning('Overwriting data exporter "{0}"'.format(exporter['name']))
+
+            self.exporters[exporter.name] = exporter
+        else:
+            logger.warning('Invalid exporter')
+
+    def isExporter(self, exporter_id):
+        if exporter_id in self.exporters:
+            return True
+        return False
+
+    def getExporters(self):
+        return self.exporters
+
+    def export(self, exporter_id, Database, PageCache):
+        data = self.exporters[exporter_id].assembler(Database, PageCache)
```

Comment:
Why bother with the intermediate result of 'data'? This could just be self.exporters[exporter_id].export(Database, PageCache). Note, this doesn't preclude assembler/formatter split in the implementations. In fact, you would just need to add
```
    def export(self, Database, PageCache):
        data = assembler(Database, PageCache)
        return formatter(data)
  ```      
 to the DataExporter class. Which if you think about it, better divides the responsibility between the classes.
 
One other thing, I would recommend not passing Database and PageCache directly as these expose 'internals'. Any changes could risk breaking exporter plugins. I would copy the information into a simple dict object and pass that, or use some sort of wrapper around them.

Reply:
I see the sense in giving the exporter an export() method, so I've done that.

Would the export manager convert the page cache to a dict? Or should that happen before the manager sees it?

I'm less clear on the right way to provide database access. The exporters run database queries directly so they only request the data needed. How does one abstract from that without having it run inefficiently or re-inventing the query system?

Reply:
Doesn't matter so much where the conversion happens, as it is easy to change it without having any knock-on effects to the export plugins. But, I think you could make a better case for it being befo.re the manager sees it based on the idea that the export manager shouldn't depend on details like DB and PageCache (I should be able to rewrite these and the export subsystem shouldnt be effected).

Your assembly methods are sort of the abstraction over the DB, so in some sense these need to be pushed up from the plugins into the manager. For instance, you could put the json assembly methods into a class and pass that in, then the exporter plugins would take the json and convert/export it. RH doesn't really have a good data abstraction layer. I guess the closest thing is the socket API which maps API calls to data queries. So you could consider passing the 'server', and the plugins can call the socket methods directly. But the server is a mess of other stuff too that would need to be re-organised.

Reply:
I think I have some understanding of the theory in what you're saying, but the specifics of implementation are a bit beyond me right now. I would be happy to revise and improve this further but I think I will need some examples to review in order to really comprehend what needs done here.